src/devices/vdml_ext_adi.c

```diff
@@ -0,0 +1,416 @@
+	struct __attribute__((packed)) {
+		double multiplier;
+		double tare_value;
+	} gyro_data;
+} adi_data_s_t;
+
+#define transform_adi_port(port)       \
+	if (port >= 'a' && port <= 'h')      \
+		port -= 'a';                       \
+	else if (port >= 'A' && port <= 'H') \
+		port -= 'A';                       \
+	else                                 \
+		port--;                            \
+	if (port > 7 || port < 0) {          \
+		errno = ENXIO;                     \
+		return PROS_ERR;                   \
+	}
+
+#define validate_type(device, port, type)                                                                 \
+	adi_port_config_e_t config = (adi_port_config_e_t)vexDeviceAdiPortConfigGet(device->device_info, port); \
+	if (config != type) {                                                                                   \
+		errno = EADDRINUSE;                                                                                   \
+		return PROS_ERR;                                                                                      \
+	}
+
+#define validate_motor(device, port)                                                                      \
+	adi_port_config_e_t config = (adi_port_config_e_t)vexDeviceAdiPortConfigGet(device->device_info, port); \
+	if (config != E_ADI_LEGACY_PWM && config != E_ADI_LEGACY_SERVO) {                                       \
+		errno = EADDRINUSE;                                                                                   \
+		return PROS_ERR;                                                                                      \
+	}
+
+	/*
+	* validate_twowire(port_top,port_bottom)
+	* 
+	* Validates 3 things:
+	* - Ports next to each other
+	* - Both ports are initialized and valid (individually)
+	* 
+	* Returns PROS_ERR if one of these is false.
+	*/
+
+#define validate_twowire(port_top, port_bottom) \
+	if (abs(port_top - port_bottom) > 1) {        \
+		errno = ENXIO;                              \
+		return PROS_ERR;                            \
+	}                                             \
+	int port;                                     \
+	if (port_top < port_bottom)                   \
+		port = port_top;                            \
+	else if (port_bottom < port_top)              \
+		port = port_bottom;                         \
+	else {                                        \
+		errno = EINVAL;                             \
+		return PROS_ERR;                            \
+	}                                             \
+	if (port % 2 == 1) {                          \
+		errno = EINVAL;                             \
+		return PROS_ERR;                            \
+	}
+
+adi_port_config_e_t ext_adi_port_get_config(uint8_t smart_port, uint8_t adi_port) {
+	transform_adi_port(adi_port);
+	claim_port_i(smart_port - 1, E_DEVICE_ADI);
+	adi_port_config_e_t rtn = (adi_port_config_e_t)vexDeviceAdiPortConfigGet(device->device_info, adi_port);
+	return_port(smart_port - 1, rtn);
+}
```

Comment:
If you have to duplicate the api, how hard would it to re-wire the existing functions to call into this one? Having two functions with the same code copy and pasted isn't that great.